### ISSUE TITLE
BProgramRunnerListener: call eventSelected on super-step only

### ISSUE THREAD

**Issue by eggsterino** [bThink-BGU/BPjs#108]
Currently, the external environment affects the bprogram state only at super-step. However, the bprogram may affect the external environment **during** a super-step as well - obviously an undesired behavior. Furthermore, it breaks the ability to roll back the program state to a previous state within the super-step (a transaction/try-catch mechanism that I am working on).

Of course, calling eventSelected on super-step must call the BProgramRunnerListeners in the same order that the events occurred.

**Comment by michbarsinai** [comment]
I think this should be implemented using an new BPJSRunner-and-verifier class. We can have many of these.

**Comment by eggsterino** [comment]
You mean - new classes that every user will create for her own project, or
in the next version of the current classes?

To be clear, I think that this is a bug and not a feature. Moreover, this
is a blocker for some smart event selection mechanisms.

**Comment by michbarsinai** [comment]
No, I mean that run-verify-run is different enough from simple running, and also a different use-case. So we can create a new class, BPjsSuperStepRunner, that will do this. I agree that it is useful, but simple running is pretty useful too :-)

**Comment by eggsterino** [comment]
run-verify-run is one option to take advantage of this change but I (and
also @geraw) think that the current situation is a misinterpretation of the
bp semantics. The whole idea of super-step is to separate the internal
logic from the environment. Currently, this separation works only in one
direction.

Fixing it will not change the semantics/traces/behavior of bp programs and
it will pave the way for run-verify-run, try-catch, undo-last-action and
other possible idioms.
